src/processing/evm.rs

```diff
@@ -0,0 +1,189 @@
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use alloy::hex::FromHex;
+use alloy::primitives::FixedBytes;
+use arc_swap::Guard;
+use bigdecimal::ToPrimitive;
+use num_bigint::BigInt;
+use starknet::core::types::{BlockId, BlockTag, Felt, FunctionCall};
+use starknet::macros::selector;
+use starknet::providers::jsonrpc::HttpTransport;
+use starknet::providers::{JsonRpcClient, Provider};
+
+use crate::config::{Config, EvmConfig};
+use crate::constants::EVM_TIME_SINCE_LAST_FEED_UPDATE;
+use crate::{config::get_config, error::MonitoringError};
+
+#[derive(Clone)]
+struct Feed {
+    feed_id: FeedId,
+    feed_type: FeedType,
+}
+
+impl TryFrom<FeedId> for Feed {
+    type Error = MonitoringError;
+
+    fn try_from(feed_id: FeedId) -> Result<Self, Self::Error>{
+        let feed_type =  FeedType::try_from(feed_id.0)?;
+        Ok(Self {
+            feed_id,
+            feed_type: feed_type,
+        })
+    }
+}
+
+impl Feed {
+    async fn get_latest_data(self, chain: &EvmConfig) -> Result<(), MonitoringError> {
+        match self.feed_type {
+            FeedType::Unique(unique_variant) => unique_variant.get_latest_data(chain, self.feed_id).await,
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+enum FeedType {
+    Unique(UniqueVariant),
+}
+
+#[derive(Clone, PartialEq, Debug)]
+enum UniqueVariant {
+    SpotMedian,
+}
+
+// let main_type = (id & 0xFF00) / 0x100;
+// let variant = id & 0x00FF;
+
+impl TryFrom<Felt> for FeedType {
+    type Error = MonitoringError;
+
+    fn try_from(feed_id: Felt) -> Result<Self, Self::Error> {
+        // Retrieving FeedIdType by dividing as bitshifting on bigint is impossible
+        let feed_id_type = feed_id.to_bigint() / BigInt::from(2).pow(216);
+        let feed_type = ((feed_id_type.clone() & BigInt::from(65280)) / BigInt::from(256))
+            .to_u64()
+            .ok_or(MonitoringError::Evm("invalid main type".to_string()))?;
+        let variant = (feed_id_type & BigInt::from(255))
+            .to_u64()
+            .ok_or(MonitoringError::Evm(format!(
+                "invalid variant for main type {}",
+                feed_type
+            )))?;
+        match feed_type {
+            0 => match variant {
+                0 => Ok(FeedType::Unique(UniqueVariant::SpotMedian)),
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+pub trait DataFetcher: Clone + Send + Sync {
+    async fn get_latest_data(&self, chain: &EvmConfig, feed_id: FeedId) -> Result<(), MonitoringError>;
+}
+
+impl DataFetcher for UniqueVariant {
+    async fn get_latest_data(&self, chain: &EvmConfig, feed_id: FeedId) -> Result<(), MonitoringError> {
```

Comment:
This should just return the return of the `call()̀  function as its name says. We get the data and we update the metric in `check_feed_update_state`

Reply:
done